File: backend/govia-identity/src/main/java/com/govia/audit/planengagement/progressreport/entity/AuditProgressReport.java
package com.govia.audit.planengagement.progressreport.entity;

import com.govia.audit.planengagement.entity.AssignmentApprovalStatus;
import com.govia.core.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;

import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * "Báo cáo tiến độ" (Khối B) - 1 chức năng TRONG màn hình "Quản lý công việc THKT" đã có (không
 * phải màn hình/permission riêng, xem AUDIT.WORK_MANAGEMENT.*). Được sinh TỰ ĐỘNG mỗi lần "Upload
 * file TTSS" (xem AuditTtssService/AuditProgressReportService.recordUpload) - không có API tạo
 * thủ công. File báo cáo đính kèm dùng bảng {@code attachment} dùng chung
 * (entityName="AUDIT_PROGRESS_REPORT", entityId=report.id), không thêm cột riêng.
 */
@Getter
@Setter
@Entity
@Table(name = "audit_progress_report")
public class AuditProgressReport extends BaseEntity {

    @Column(name = "engagement_id", nullable = false, columnDefinition = "uuid")
    private UUID engagementId;

    /** "mã mảng nv" - tham chieu AuditMasterDataItem category BUSINESS_SEGMENT. */
    @Column(name = "business_segment_id", columnDefinition = "uuid")
    private UUID businessSegmentId;

    @Column(name = "reported_employee_id", columnDefinition = "uuid")
    private UUID reportedEmployeeId;

    @Column(name = "total_findings", nullable = false)
    private int totalFindings;

    @Column(name = "total_ttss", nullable = false)
    private int totalTtss;

    @Column(name = "total_material_findings", nullable = false)
    private int totalMaterialFindings;

    @Column(name = "total_material_ttss", nullable = false)
    private int totalMaterialTtss;

    /** Tổng số lượng mẫu - proxy = đếm AuditEngagementAssignment theo user+mảng NV (không đếm mẫu
     * thật ở 16 bảng chọn mẫu CmNtd/CmTd - xem ghi chú trong plan). */
    @Column(name = "total_samples", nullable = false)
    private int totalSamples;

    @Column(name = "completed_samples", nullable = false)
    private int completedSamples;

    @Column(name = "report_date", nullable = false)
    private LocalDate reportDate;

    /** "Lần báo cáo" - tăng dần theo (engagementId, businessSegmentId, reportedEmployeeId). */
    @Column(name = "report_round", nullable = false)
    private int reportRound;

    @Column(name = "reported_by_username", length = 100)
    private String reportedByUsername;

    @Column(name = "note", length = 2000)
    private String note;

    @Enumerated(EnumType.STRING)
    @Column(name = "approval_status", length = 20)
    private AssignmentApprovalStatus approvalStatus;

    @Column(name = "approved_by", length = 100)
    private String approvedBy;

    @Column(name = "approved_at")
    private Instant approvedAt;

    @Column(name = "process_instance_id", length = 64)
    private String processInstanceId;
}

File: frontend/apps/shell/src/api/auditProgressReport.ts
import axios from "axios";
import type { ApiResponse } from "@govia/ui-kit";
import { httpClient } from "./client";
import type { AssignmentApprovalStatus } from "./auditWorkManagement";

export interface AuditProgressReportItem {
  id: string;
  engagementId: string;
  businessSegmentId: string | null;
  businessSegmentCode: string | null;
  reportedEmployeeId: string | null;
  reportedEmployeeCode: string | null;
  reportedEmployeeName: string | null;
  totalFindings: number;
  totalTtss: number;
  totalMaterialFindings: number;
  totalMaterialTtss: number;
  totalSamples: number;
  completedSamples: number;
  reportDate: string;
  reportRound: number;
  reportedByUsername: string | null;
  note: string | null;
  approvalStatus: AssignmentApprovalStatus | null;
  approvedBy: string | null;
  approvedAt: string | null;
}

export interface AuditProgressReportAttachment {
  id: string;
  fileName: string;
}

function base(engagementId: string): string {
  return `/api/audit/plan/engagement/${engagementId}/work-management/progress-reports`;
}

export async function listAuditProgressReports(engagementId: string): Promise<AuditProgressReportItem[]> {
  const res = await httpClient.get<ApiResponse<AuditProgressReportItem[]>>(base(engagementId));
  return res.data.data;
}

export async function approveAuditProgressReports(engagementId: string, reportIds: string[]): Promise<string[]> {
  const res = await httpClient.post<ApiResponse<string[]>>(`${base(engagementId)}/approve`, { reportIds });
  return res.data.data;
}

/** File TTSS goc dinh kem tu dong luc sinh dong bao cao (xem AuditProgressReportService.recordUpload
 * - luu qua AttachmentService dung chung, entityName="AUDIT_PROGRESS_REPORT", entityId=report.id). */
export async function listAuditProgressReportAttachments(reportId: string): Promise<AuditProgressReportAttachment[]> {
  const res = await httpClient.get<ApiResponse<AuditProgressReportAttachment[]>>("/api/attachments", {
    params: { entityName: "AUDIT_PROGRESS_REPORT", entityId: reportId },
  });
  return res.data.data;
}

export async function downloadAuditProgressReportAttachment(attachmentId: string, fileName: string): Promise<void> {
  try {
    const res = await httpClient.get(`/api/attachments/${attachmentId}/download`, { responseType: "blob" });
    const blobUrl = window.URL.createObjectURL(res.data as Blob);
    const link = document.createElement("a");
    link.href = blobUrl;
    link.download = fileName;
    link.click();
    window.URL.revokeObjectURL(blobUrl);
  } catch (err) {
    // responseType "blob" ap dung ca cho response LOI - err.response.data la 1 Blob (chua JSON dang
    // text) thay vi object da parse, nen getApiErrorMessage() luon fallback ve thong bao chung chung
    // du backend tra ve message cu the (vd "File dinh kem khong con tren dia..."). Doc lai thanh JSON
    // truoc khi nem tiep de UI hien dung ly do that su.
    if (axios.isAxiosError(err) && err.response?.data instanceof Blob) {
      try {
        err.response.data = JSON.parse(await err.response.data.text());
      } catch {
        // khong parse duoc (vd loi mang tra ve HTML) - giu nguyen blob, getApiErrorMessage fallback binh thuong
      }
    }
    throw err;
  }
}
